File: ios/Classes/ComTappSoAdjustPlugin.swift
import Flutter
import UIKit
import TappAdjust
import Tapp

public class ComTappSoAdjustPlugin: NSObject, FlutterPlugin {

  private var methodChannel: FlutterMethodChannel!
  private var eventChannel: FlutterEventChannel!
  private var eventSink: FlutterEventSink?

  // MARK: - Registration

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = ComTappSoAdjustPlugin()

    // Method channel
    instance.methodChannel = FlutterMethodChannel(
      name: "com.tapp.so.adjust/methods",
      binaryMessenger: registrar.messenger()
    )
    registrar.addMethodCallDelegate(instance, channel: instance.methodChannel)

    // Event channel for deferred-link callbacks
    instance.eventChannel = FlutterEventChannel(
      name: "com.tapp.so.adjust/events",
      binaryMessenger: registrar.messenger()
    )
    instance.eventChannel.setStreamHandler(instance)
  }

  // MARK: - MethodChannel

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    let args = call.arguments as? [String: Any]

    switch call.method {
    // ---- Core SDK ----

    case "initialize":
      guard
        let auth      = args?["authToken"]   as? String,
        let envStr    = args?["environment"] as? String,
        let tappToken = args?["tappToken"]   as? String
      else {
        result(FlutterError(code: "BAD_ARGS", message: "Missing initialize args", details: nil))
        return
      }
      let bundleID = Bundle.main.bundleIdentifier ?? ""
      let config = TappConfiguration(
        authToken: auth,
        env: envStr,                  // string-based initializer (e.g., "PRODUCTION"/"SANDBOX")
        tappToken: tappToken,
        affiliateName: "adjust",     // e.g., "adjust"
        bundleID: bundleID
      )
      Tapp.start(config: config, delegate: self)
      result(nil)

    case "getPlatformVersion":
      result("iOS " + UIDevice.current.systemVersion)

    case "shouldProcess":
      guard
        let link = args?["deepLink"] as? String,
        let url  = URL(string: link)
      else {
        result(FlutterError(code: "BAD_ARGS", message: "Expected deepLink string", details: nil))
        return
      }
      result(Tapp.shouldProcess(url: url))

    case "generateUrl":
      guard let influencer = args?["influencer"] as? String else {
        result(FlutterError(code: "BAD_ARGS", message: "Missing influencer", details: nil))
        return
      }
      let adGroup  = args?["adGroup"]  as? String
      let creative = args?["creative"] as? String
      let data     = args?["data"]     as? [String: String]
      let urlConfig = AffiliateURLConfiguration(
        influencer: influencer,
        adgroup: adGroup,
        creative: creative,
        data: data
      )
      Tapp.url(config: urlConfig) { response, error in
        if let err = error {
          result(FlutterError(code: "URL_ERROR", message: err.localizedDescription, details: nil))
        } else if let resp = response {
          result(resp.url.absoluteString)
        } else {
          result(FlutterError(code: "URL_ERROR", message: "Unexpected nil response", details: nil))
        }
      }

    case "handleEvent":
      guard let token = args?["eventToken"] as? String else {
        result(FlutterError(code: "BAD_ARGS", message: "Missing eventToken", details: nil))
        return
      }
      let cfg = EventConfig(eventToken: token)
      Tapp.handleEvent(config: cfg)
      result(nil)

    case "handleTappEvent":
      guard let action = args?["eventAction"] as? String else {
        result(FlutterError(code: "BAD_ARGS", message: "Missing eventAction", details: nil))
        return
      }
      let event = TappEvent(eventActionName: action)
      Tapp.handleTappEvent(event: event)
      result(nil)

    case "fetchLinkData":
      guard
        let link = args?["deepLink"] as? String,
        let url  = URL(string: link)
      else {
        result(FlutterError(code: "BAD_ARGS", message: "Expected deepLink", details: nil))
        return
      }
      Tapp.fetchLinkData(for: url) { data, error in
        if let err = error {
          result([
            "error": true,
            "message": err.localizedDescription
          ])
        } else if let d = data {
          result([
            "error": false,
            "message": "",
            "tappUrl": d.tappURL.absoluteString,
            "attrTappUrl": d.attributedTappURL.absoluteString,
            "influencer": d.influencer,
            "isFirstSession": d.isFirstSession,
            "data": d.data ?? [:]
          ])
        } else {
          result([
            "error": true,
            "message": "Unknown error"
          ])
        }
      }

    case "fetchOriginLinkData":
      Tapp.fetchOriginLinkData { data, error in
        if let err = error {
          result([
            "error": true,
            "message": err.localizedDescription
          ])
        } else if let d = data {
          result([
            "error": false,
            "message": "",
            "tappUrl": d.tappURL.absoluteString,
            "attrTappUrl": d.attributedTappURL.absoluteString,
            "influencer": d.influencer,
            "isFirstSession": d.isFirstSession,
            "data": d.data ?? [:]
          ])
        } else {
          result([
            "error": true,
            "message": "Unknown error"
          ])
        }
      }

    case "getConfig":
      // iOS: not supported
      print("[com_tapp_so_adjust] getConfig not supported on iOS")
      result([
        "error": true,
        "message": "getConfig not supported on iOS"
      ])

    // ---- Adjust integration ----

    case "adjustEnable":
      Tapp.adjustEnable(); result(nil)

    case "adjustDisable":
      Tapp.adjustDisable(); result(nil)

    case "adjustIsEnabled":
      Tapp.adjustIsEnabled { num in result(num.boolValue) }

    case "adjustGdprForgetMe":
      Tapp.adjustGdprForgetMe(); result(nil)

    case "adjustSetReferrer":
      // Not supported on iOS
      print("[com_tapp_so_adjust] adjustSetReferrer not supported on iOS")
      result([
        "error": true,
        "message": "adjustSetReferrer not supported on iOS"
      ])

    case "adjustSetPushToken":
      // Not supported on iOS
      print("[com_tapp_so_adjust] adjustSetPushToken not supported on iOS")
      result([
        "error": true,
        "message": "adjustSetPushToken not supported on iOS"
      ])

    case "adjustTrackAdRevenue":
      if
        let src = args?["source"]  as? String,
        let rev = args?["revenue"] as? Double,
        let cur = args?["currency"] as? String
      {
        Tapp.adjustTrackAdRevenue(source: src, revenue: rev, currency: cur)
      }
      result(nil)

    case "adjustTrackThirdPartySharing":
      if let en = args?["enabled"] as? Bool {
        Tapp.adjustTrackThirdPartySharing(isEnabled: en)
      }
      result(nil)

    case "getAdjustAttribution":
      Tapp.getAdjustAttribution { attr in
        var map: [String:Any] = [:]
        if let a = attr {
          map = [
            "trackerToken": a.trackerToken ?? "",
            "trackerName":  a.trackerName  ?? "",
            "network":      a.network      ?? "",
            "campaign":     a.campaign     ?? "",
            "adgroup":      a.adGroup      ?? "",
            "creative":     a.creative     ?? "",
            "clickLabel":   a.clickLabel   ?? "",
            "costType":     a.costType     ?? "",
            "costAmount":   a.costAmount?.doubleValue ?? 0.0,
            "costCurrency": a.costCurrency ?? ""
          ]
          if let dict = a.dictionary as? [String:Any] {
            map["fbInstallReferrer"] = dict["fbInstallReferrer"] as? String ?? ""
            map["costInUsd"]         = dict["costInUsd"]         as? Double ?? 0.0
            map["callbackParams"]    = dict["callbackParams"]    as? [String:Any] ?? [:]
            map["partnerParams"]     = dict["partnerParams"]     as? [String:Any] ?? [:]
          }
        }
        result(["attribution": map])
      }

    case "adjustVerifyAppStorePurchase":
      guard
        let tx  = args?["transactionId"] as? String,
        let pid = args?["productId"]     as? String
      else {
        result(FlutterError(code: "BAD_ARGS", message: "Missing transactionId or productId", details: nil))
        return
      }
      Tapp.adjustVerifyAppStorePurchase(transactionId: tx, productId: pid) { resOpt in
        guard let res = resOpt else {
          result(FlutterError(code: "VERIFY_ERROR", message: "No result returned", details: nil))
          return
        }
        result([
          "verificationStatus": res.verificationStatus,
          "code":               Int(res.code),
          "message":            res.message
        ])
      }

    case "adjustTrackAppStoreSubscription":
      if
        let sub = args?["subscription"]      as? [String: Any],
        let price = sub["price"]             as? Double,
        let cur = sub["currency"]            as? String,
        let tx = sub["transactionId"]        as? String,
        let subscription = AdjustAppStoreSubscription(
          price: NSDecimalNumber(value: price),
          currency: cur,
          transactionId: tx
        )
      {
        Tapp.adjustTrackAppStoreSubscription(subscription)
        result(nil)
      } else {
        result(FlutterError(code: "BAD_ARGS", message: "Invalid subscription parameters", details: nil))
      }

    case "adjustConvert":
      guard
        let link = args?["universalLink"] as? String,
        let scheme = args?["scheme"] as? String,
        let url = URL(string: link)
      else {
        result(FlutterError(code: "BAD_ARGS", message: "Invalid convert args", details: nil))
        return
      }
      let convertedURL = Tapp.adjustConvert(universalLink: url, with: scheme)
      result(convertedURL?.absoluteString)

    case "adjustRequestAppTrackingAuthorization":
      Tapp.adjustRequestAppTrackingAuthorization { status in result(status) }

    case "adjustAppTrackingAuthorizationStatus":
      result(Tapp.adjustAppTrackingAuthorizationStatus())

    case "adjustUpdateSkanConversionValue":
      guard
        let p   = args?["params"]    as? [String:Any],
        let val = p["value"]         as? Int,
        let coarse = p["coarseValue"] as? String
      else {
        result(FlutterError(code: "BAD_ARGS", message: "Missing SKAN conversion parameters", details: nil))
        return
      }
      let lockNumber: NSNumber? = (p["lockWindow"] as? Int).map { NSNumber(value: $0) }
      Tapp.adjustUpdateSkanConversionValue(val, coarseValue: coarse, lockWindow: lockNumber) { err in
        if let e = err {
          result(FlutterError(code: "SKAN_ERROR", message: e.localizedDescription, details: nil))
        } else {
          result(nil)
        }
      }

    default:
      result(FlutterMethodNotImplemented)
    }
  }
}

// MARK: - EventChannel

extension ComTappSoAdjustPlugin: FlutterStreamHandler {
  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    self.eventSink = events
    return nil
  }
  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    self.eventSink = nil
    return nil
  }
}

// MARK: - TappDelegate

extension ComTappSoAdjustPlugin: TappDelegate {
  // Called when a deferred deep link is opened
  public func didOpenApplication(with data: TappDeferredLinkData) {
    guard let sink = self.eventSink else { return }
    let map: [String:Any] = [
      "type": "onDeferredLinkReceived",
      "error": false,
      "message": "",
      "tappUrl": data.tappURL.absoluteString,
      "attrTappUrl": data.attributedTappURL.absoluteString,
      "influencer": data.influencer,
      "isFirstSession": data.isFirstSession,
      "data": data.data ?? [:]
    ]
    sink(map)
  }

  // Called when resolving the deferred URL fails
  public func didFailResolvingURL(url: URL, error: Error) {
    guard let sink = self.eventSink else { return }
    let map: [String:Any] = [
      "type": "onDidFailResolvingURL",
      "url": url.absoluteString,
      "error": error.localizedDescription
    ]
    sink(map)
  }
}

File: android/src/main/kotlin/com/tapp/so/adjust/ComTappSoAdjustPlugin.kt
package com.tapp.so.adjust

import android.content.Context
import android.util.Log
import com.adjust.sdk.AdjustEvent
import com.adjust.sdk.AdjustPurchaseVerificationResult
import com.adjust.sdk.GooglePlayInstallReferrerDetails
import com.adjust.sdk.OnGooglePlayInstallReferrerReadListener
import com.adjust.sdk.OnIsEnabledListener
import com.adjust.sdk.OnPurchaseVerificationFinishedListener
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.*

// ------- Tapp SDK (your native lib) -------
import com.example.tapp.Tapp
import com.example.tapp.models.Environment
import com.example.tapp.models.Affiliate
import com.example.tapp.services.network.RequestModels
import com.example.tapp.services.affiliate.tapp.DeferredLinkDelegate

/** ComTappSoAdjustPlugin */
class ComTappSoAdjustPlugin : FlutterPlugin, MethodChannel.MethodCallHandler {

  companion object {
    private const val TAG = "ComTappSoAdjustPlugin"
    private const val METHOD_CHANNEL = "com.tapp.so.adjust/methods"
    private const val EVENT_CHANNEL  = "com.tapp.so.adjust/events"
  }

  private lateinit var methodChannel: MethodChannel
  private lateinit var eventChannel: EventChannel
  private lateinit var context: Context

  // Single Tapp instance
  private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
  private var eventSink: EventChannel.EventSink? = null
  private lateinit var tapp: Tapp

  override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    Log.d(TAG, "onAttachedToEngine")
    context = binding.applicationContext
    tapp = Tapp(context)

    methodChannel = MethodChannel(binding.binaryMessenger, METHOD_CHANNEL)
    methodChannel.setMethodCallHandler(this)

    eventChannel = EventChannel(binding.binaryMessenger, EVENT_CHANNEL)
    eventChannel.setStreamHandler(object : EventChannel.StreamHandler {
      override fun onListen(args: Any?, sink: EventChannel.EventSink) {
        Log.d(TAG, "EventChannel.onListen args=$args")
        eventSink = sink

        // Wire native deferred link callbacks to Flutter stream
        tapp.deferredLinkDelegate = object : DeferredLinkDelegate {
          override fun didReceiveDeferredLink(resp: RequestModels.TappLinkDataResponse) {
            Log.d(TAG, "didReceiveDeferredLink → $resp")
            sink.success(
              mapOf(
                "type"           to "onDeferredLinkReceived",
                "error"          to resp.error,
                "message"        to resp.message,
                "tappUrl"        to resp.tappUrl,
                "attrTappUrl"    to resp.attrTappUrl,
                "influencer"     to resp.influencer,
                "isFirstSession" to (resp.isFirstSession ?: false),
                "data"           to (resp.data ?: emptyMap<String, String>())
              )
            )
          }

          override fun didFailResolvingUrl(resp: RequestModels.FailResolvingUrlResponse) {
            Log.d(TAG, "didFailResolvingUrl → $resp")
            sink.success(
              mapOf(
                "type"  to "onDidFailResolvingURL",
                "url"   to resp.url,
                "error" to resp.error
              )
            )
          }

          override fun testListener(test: String) {
            sink.success(mapOf("type" to "onTestListener", "test" to test))
          }
        }
      }

      override fun onCancel(args: Any?) {
        Log.d(TAG, "EventChannel.onCancel args=$args")
        eventSink = null
        tapp.deferredLinkDelegate = null
      }
    })
  }

  override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
    Log.d(TAG, "onMethodCall: ${call.method} ← ${call.arguments}")
    when (call.method) {
      "initialize"          -> initialize(call, result)
      "getPlatformVersion"  -> result.success("Android ${android.os.Build.VERSION.RELEASE}")
      "shouldProcess"       -> {
        val url = call.argument<String>("deepLink")
        Log.d(TAG, "shouldProcess(deepLink=$url)")
        result.success(tapp.shouldProcess(url))
      }
      "generateUrl"         -> generateUrl(call, result)
      "handleEvent"         -> {
        Log.d(TAG, "handleEvent(token=${call.argument<String>("eventToken")})")
        tapp.handleEvent(call.argument<String>("eventToken")!!)
        result.success(null)
      }
      "handleTappEvent"     -> handleTappEvent(call, result)
      "fetchLinkData"       -> fetchLinkData(call, result)
      "fetchOriginLinkData" -> fetchOriginLinkData(result)
      "getConfig"           -> {
        Log.d(TAG, "getConfig()")
        val response = tapp.getConfig()
        Log.d(TAG, "getConfig raw → $response")
        result.success(response.toMap())
      }
      "simulateTestEvent"   -> { tapp.simulateTestEvent(); result.success(null) }

      // —— Adjust integration ——
      "adjustEnable"         -> { tapp.adjustEnable(); result.success(null) }
      "adjustDisable"        -> { tapp.adjustDisable(); result.success(null) }
      "adjustIsEnabled"      -> adjustIsEnabled(result)
      "adjustGdprForgetMe"   -> { tapp.adjustGdprForgetMe(); result.success(null) }
      "adjustSetReferrer"    -> { tapp.adjustSetReferrer(call.argument<String>("referrer")!!); result.success(null) }
      "adjustSetPushToken"   -> { tapp.adjustSetPushToken(call.argument<String>("token")!!); result.success(null) }
      "adjustOnResume"       -> { tapp.adjustOnResume(); result.success(null) }
      "adjustOnPause"        -> { tapp.adjustOnPause(); result.success(null) }

      "adjustTrackAdRevenue"          -> trackAdRevenue(call, result)
      "adjustTrackThirdPartySharing"  -> { tapp.adjustTrackThirdPartySharing(call.argument<Boolean>("enabled")!!); result.success(null) }
      "adjustTrackMeasurementConsent" -> { tapp.adjustTrackMeasurementConsent(call.argument<Boolean>("consent")!!); result.success(null) }

      "adjustAddGlobalCallbackParameter"     -> { tapp.adjustAddGlobalCallbackParameter(call.argument<String>("key")!!, call.argument<String>("value")!!); result.success(null) }
      "adjustAddGlobalPartnerParameter"      -> { tapp.adjustAddGlobalPartnerParameter(call.argument<String>("key")!!, call.argument<String>("value")!!); result.success(null) }
      "adjustRemoveGlobalCallbackParameter"  -> { tapp.adjustRemoveGlobalCallbackParameter(call.argument<String>("key")!!); result.success(null) }
      "adjustRemoveGlobalPartnerParameter"   -> { tapp.adjustRemoveGlobalPartnerParameter(call.argument<String>("key")!!); result.success(null) }
      "adjustRemoveGlobalCallbackParameters" -> { tapp.adjustRemoveGlobalCallbackParameters(); result.success(null) }
      "adjustRemoveGlobalPartnerParameters"  -> { tapp.adjustRemoveGlobalPartnerParameters(); result.success(null) }

      "adjustVerifyAppStorePurchase"          -> verifyAppStorePurchase(call, result)
      "adjustVerifyAndTrackPlayStorePurchase" -> adjustVerifyAndTrackPlayStorePurchase(call, result)
      "adjustTrackPlayStoreSubscription"      -> trackPlayStoreSubscription(call, result)

      "adjustGetAdid"                      -> getAsync { tapp.adjustGetAdid(it) }.toPromise(result)
      "adjustGetIdfa"                      -> getAsync { tapp.adjustGetIdfa(it) }.toPromise(result)
      "adjustGetGoogleAdId"                -> getAsync { tapp.adjustGetGoogleAdId(it) }.toPromise(result)
      "adjustGetAmazonAdId"                -> getAsync { tapp.adjustGetAmazonAdId(it) }.toPromise(result)
      "adjustGetSdkVersion"                -> getAsync { tapp.adjustGetSdkVersion(it) }.toPromise(result)
      "adjustGetGooglePlayInstallReferrer" -> getReferrer(result)

      else -> {
        Log.w(TAG, "Method not implemented: ${call.method}")
        result.notImplemented()
      }
    }
  }

  // — Helpers —

  private fun initialize(call: MethodCall, result: MethodChannel.Result) {
    Log.d(TAG, "initialize() args=${call.arguments}")
    val args      = call.arguments as? Map<*, *> ?: emptyMap<Any, Any>()
    val auth      = args["authToken"]   as? String ?: ""
    val envStr    = args["environment"] as? String ?: "SANDBOX"
    val tappToken = args["tappToken"]   as? String ?: ""

    val env = if (envStr.equals("PRODUCTION", true) || envStr.equals("PROD", true))
      Environment.PRODUCTION else Environment.SANDBOX

    val aff = Affiliate.ADJUST

    tapp.start(com.example.tapp.utils.TappConfiguration(auth, env, tappToken, aff))
    Log.d(TAG, "initialize() complete")
    result.success(null)
  }

  private fun generateUrl(call: MethodCall, result: MethodChannel.Result) {
    Log.d(TAG, "generateUrl() args=${call.arguments}")
    val influencer = call.argument<String>("influencer")
      ?: return result.error("BAD_ARGS", "Missing influencer", null)
    val adGroup  = call.argument<String>("adGroup")
    val creative = call.argument<String>("creative")
    val data     = (call.argument<Map<String, String>>("data") ?: emptyMap())

    scope.launch {
      try {
        val resp = tapp.url(influencer, adGroup, creative, data)
        Log.d(TAG, "generateUrl() resp=$resp")
        withContext(Dispatchers.Main) { result.success(resp.influencer_url) }
      } catch (e: Exception) {
        Log.e(TAG, "generateUrl() error", e)
        withContext(Dispatchers.Main) { result.error("URL_ERROR", e.message, null) }
      }
    }
  }

  private fun handleTappEvent(call: MethodCall, result: MethodChannel.Result) {
    Log.d(TAG, "handleTappEvent() args=${call.arguments}")
    val actionStr = call.argument<String>("eventAction")?.lowercase()
      ?: return result.error("BAD_ARGS", "Missing eventAction", null)

    scope.launch {
      tapp.handleTappEvent(RequestModels.TappEvent(RequestModels.EventAction.custom(actionStr)))
      Log.d(TAG, "handleTappEvent() complete")
      withContext(Dispatchers.Main) { result.success(null) }
    }
  }

  private fun fetchLinkData(call: MethodCall, result: MethodChannel.Result) {
    Log.d(TAG, "fetchLinkData() args=${call.arguments}")
    val url = call.argument<String>("deepLink")
      ?: return result.error("BAD_ARGS", "Missing URL", null)

    scope.launch {
      val resp = tapp.fetchLinkData(url)
      Log.d(TAG, "fetchLinkData() resp=$resp")
      val map = resp?.let {
        mapOf(
          "error"          to it.error,
          "message"        to it.message,
          "tappUrl"        to it.tappUrl,
          "attrTappUrl"    to it.attrTappUrl,
          "influencer"     to it.influencer,
          "isFirstSession" to (it.isFirstSession ?: false),
          "data"           to (it.data ?: emptyMap<String, String>())
        )
      } ?: mapOf("error" to true, "message" to "no response")
      withContext(Dispatchers.Main) { result.success(map) }
    }
  }

  private fun fetchOriginLinkData(result: MethodChannel.Result) {
    Log.d(TAG, "fetchOriginLinkData()")
    scope.launch {
      val resp = tapp.fetchOriginalLinkData()
      Log.d(TAG, "fetchOriginLinkData() resp=$resp")
      val map = resp?.let {
        mapOf(
          "error"          to it.error,
          "message"        to it.message,
          "tappUrl"        to it.tappUrl,
          "attrTappUrl"    to it.attrTappUrl,
          "influencer"     to it.influencer,
          "isFirstSession" to (it.isFirstSession ?: false),
          "data"           to (it.data ?: emptyMap<String, String>())
        )
      } ?: mapOf("error" to true, "message" to "no response")
      withContext(Dispatchers.Main) { result.success(map) }
    }
  }

  private fun trackAdRevenue(call: MethodCall, result: MethodChannel.Result) {
    Log.d(TAG, "trackAdRevenue() args=${call.arguments}")
    tapp.adjustTrackAdRevenue(
      call.argument<String>("source")!!,
      call.argument<Double>("revenue")!!,
      call.argument<String>("currency")!!
    )
    result.success(null)
  }

  private fun adjustIsEnabled(result: MethodChannel.Result) {
    Log.d(TAG, "adjustIsEnabled()")
    tapp.adjustIsEnabled(context, object : OnIsEnabledListener {
      override fun onIsEnabledRead(isEnabled: Boolean) {
        Log.d(TAG, "adjustIsEnabled callback → $isEnabled")
        result.success(isEnabled)
      }
    })
  }

  private fun verifyAppStorePurchase(call: MethodCall, result: MethodChannel.Result) {
    Log.d(TAG, "verifyAppStorePurchase() args=${call.arguments}")
    val txId = call.argument<String>("transactionId")!!
    val pid  = call.argument<String>("productId")!!
    tapp.adjustVerifyAppStorePurchase(txId, pid) { res ->
      Log.d(TAG, "verifyAppStorePurchase callback → $res")
      result.success(
        mapOf(
          "verificationStatus" to res.verificationStatus,
          "code"               to res.code,
          "message"            to res.message
        )
      )
    }
  }

  private fun adjustVerifyAndTrackPlayStorePurchase(call: MethodCall, result: MethodChannel.Result) {
    Log.d(TAG, "adjustVerifyAndTrackPlayStorePurchase() args=${call.arguments}")
    val token = call.argument<String>("eventToken")!!
    val event = AdjustEvent(token)
    tapp.adjustVerifyAndTrackPlayStorePurchase(event, object : OnPurchaseVerificationFinishedListener {
      override fun onVerificationFinished(res: AdjustPurchaseVerificationResult) {
        Log.d(TAG, "adjustVerifyAndTrackPlayStorePurchase callback → $res")
        result.success(
          mapOf(
            "verificationStatus" to res.verificationStatus,
            "code"               to res.code,
            "message"            to res.message
          )
        )
      }
    })
  }

  private fun trackPlayStoreSubscription(call: MethodCall, result: MethodChannel.Result) {
    Log.d(TAG, "trackPlayStoreSubscription() args=${call.arguments}")
    val sub = call.argument<Map<String, Any>>("subscription")!!
    tapp.adjustTrackPlayStoreSubscription(
      sub["price"] as Long,
      sub["currency"] as String,
      sub["sku"] as String,
      sub["orderId"] as String,
      sub["signature"] as String,
      sub["purchaseToken"] as String,
      sub["purchaseTime"] as? Long
    )
    result.success(null)
  }

  private fun getReferrer(result: MethodChannel.Result) {
    Log.d(TAG, "getReferrer()")
    tapp.adjustGetGooglePlayInstallReferrer(object : OnGooglePlayInstallReferrerReadListener {
      override fun onInstallReferrerRead(details: GooglePlayInstallReferrerDetails?) {
        Log.d(TAG, "getReferrer callback → $details")
        result.success(details?.toString() ?: "")
      }

      override fun onFail(msg: String?) {
        Log.e(TAG, "getReferrer error → $msg")
        result.error("INSTALL_REFERRER_ERROR", msg, null)
      }
    })
  }

  // Utility to wrap async callbacks into a one-shot promise
  private fun <T> getAsync(callee: ((T?) -> Unit) -> Unit) = object {
    fun toPromise(result: MethodChannel.Result) {
      callee.invoke { v ->
        if (v != null) result.success(v) else result.error("NULL", "No value", null)
      }
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    Log.d(TAG, "onDetachedFromEngine")
    scope.cancel()
    methodChannel.setMethodCallHandler(null)
    eventChannel.setStreamHandler(null)
  }

  // Convert ConfigResponse → Map
  private fun RequestModels.ConfigResponse.toMap() = mapOf(
    "error"   to this.error,
    "message" to this.message,
    "config"  to (this.config?.let {
      mapOf(
        "authToken"   to it.authToken,
        "env"         to it.env.name,
        "tappToken"   to it.tappToken,
        "affiliate"   to it.affiliate.name,
        "bundleID"    to (it.bundleID ?: ""),
        "appToken"    to (it.appToken ?: ""),
        "deepLinkUrl" to (it.deepLinkUrl ?: ""),
        "linkToken"   to (it.linkToken ?: "")
      )
    } ?: emptyMap<String, Any>())
  )
}
